From 8cebd4ffe67bf94508809ea0caa02a4f1d52e8b1 Mon Sep 17 00:00:00 2001
From: Carlos O'Donell <carlos@redhat.com>
Date: Mon, 26 Nov 2018 09:51:51 -0500
Subject: [PATCH] Add --no-hard-links option to localedef (bug 23923)

Downstream distributions need consistent sets of hardlinks in
order for rpm to operate effectively. This means that even if
locales are built with a high level of parallelism that the
resulting files need to have consistent hardlink counts. The only
way to achieve this is with a post-install hardlink pass using a
program like 'hardlink' (shipped in Fedora).

If the downstream distro wants to post-process the hardlinks then
the time spent in localedef looking up sibling directories and
processing hardlinks is wasted effort.

To optimize the build and install pass we add a --no-hard-links
option to localedef to avoid doing the hardlink optimziation for
size.

Tested on x86_64 with 'make localedata/install-locale-files'
before and after. Without the patch we have files with 100+
hardlink counts. After the patch and running with --no-hard-links
all link counts are 1. This patch also alters the convenience
target 'make localedata/install-locale-files' to use the new
option.

Signed-off-by: Carlos O'Donell <carlos@redhat.com>
---
 ChangeLog                   |   10 ++++++++++
 locale/programs/localedef.c |   11 +++++++++++
 locale/programs/localedef.h |    1 +
 locale/programs/locfile.c   |   21 +++++++++++++++++----
 localedata/Makefile         |    2 +-
 5 files changed, 40 insertions(+), 5 deletions(-)

--- a/ChangeLog
+++ b/ChangeLog
@@ -24393,4 +24393,14 @@
 	* sysdeps/i386/i686/multiarch/memcpy-ssse3.S (MEMCPY_CHK):
 	Likewise.
 
+2018-12-03  Carlos O'Donell  <carlos@redhat.com>
+
+	[BZ #23923]
+	* locale/programs/localedef.c: Declare boolean hard_links default true.
+	(options): Add --no-hard-links option.
+	(parse_opt): Add OPT_NO_HARD_LINKS case and set hard_links to false.
+	* locale/programs/localedef.h: Declare prototype for hard_links.
+	* locale/programs/locfile.c (write_locale_data): Don't use hard
+	links if hard_links is false.
+
 See ChangeLog.18 for earlier changes.
--- a/locale/programs/localedef.c
+++ b/locale/programs/localedef.c
@@ -85,6 +85,9 @@ static bool replace_archive;
 /* If true list archive content.  */
 static bool list_archive;
 
+/* If true create hard links to other locales (default).  */
+bool hard_links = true;
+
 /* Maximum number of retries when opening the locale archive.  */
 int max_locarchive_open_retry = 10;
 
@@ -106,6 +109,8 @@ void (*argp_program_version_hook) (FILE
 #define OPT_NO_WARN 402
 #define OPT_WARN 403
 #define OPT_UINT32_ALIGN 404
+#define OPT_NO_HARD_LINKS 405
+
 
 /* Definitions of arguments for argp functions.  */
 static const struct argp_option options[] =
@@ -121,6 +126,8 @@ static const struct argp_option options[
   { NULL, 0, NULL, 0, N_("Output control:") },
   { "force", 'c', NULL, 0,
     N_("Create output even if warning messages were issued") },
+  { "no-hard-links", OPT_NO_HARD_LINKS, NULL, 0,
+    N_("Do not create hard links between installed locales") },
   { "prefix", OPT_PREFIX, N_("PATH"), 0, N_("Optional output file prefix") },
   { "posix", OPT_POSIX, NULL, 0, N_("Strictly conform to POSIX") },
   { "quiet", OPT_QUIET, NULL, 0,
@@ -397,6 +404,10 @@ parse_opt (int key, char *arg, struct ar
     case OPT_UINT32_ALIGN:
       uint32_align_mask = strtol (arg, NULL, 0) - 1;
       break;
+    case OPT_NO_HARD_LINKS:
+      /* Do not hard link to other locales.  */
+      hard_links = false;
+      break;
     case 'c':
       force_output = 1;
       break;
--- a/locale/programs/localedef.h
+++ b/locale/programs/localedef.h
@@ -118,6 +118,7 @@ extern const char *repertoire_global;
 extern int max_locarchive_open_retry;
 extern bool no_archive;
 extern const char *alias_file;
+extern bool hard_links;
 
 
 /* Prototypes for a few program-wide used functions.  */
--- a/locale/programs/locfile.c
+++ b/locale/programs/locfile.c
@@ -705,7 +705,7 @@ write_locale_data (const char *output_pa
   size_t cnt, step, maxiov;
   int fd;
   char *fname;
-  const char **other_paths;
+  const char **other_paths = NULL;
   uint32_t header[2];
   size_t n_elem;
   struct iovec vec[3];
@@ -830,9 +830,22 @@ failure while writing data for category
 
   close (fd);
 
-  /* Compare the file with the locale data files for the same category in
-     other locales, and see if we can reuse it, to save disk space.  */
-  other_paths = siblings (output_path);
+  /* Compare the file with the locale data files for the same category
+     in other locales, and see if we can reuse it, to save disk space.
+     If the user specified --no-hard-links to localedef then hard_links
+     is false, other_paths remains NULL and we skip the optimization
+     below.  The use of --no-hard-links is distribution specific since
+     some distros have post-processing hard-link steps and so doing this
+     here is a waste of time.  Worse than a waste of time in rpm-based
+     distributions it can result in build determinism issues from
+     build-to-build since some files may get a hard link in one pass but
+     not in another (if the files happened to be created in parallel).  */
+  if (hard_links)
+    other_paths = siblings (output_path);
+
+  /* If there are other paths, then walk the sibling paths looking for
+     files with the same content so we can hard link and reduce disk
+     space usage.  */
   if (other_paths != NULL)
     {
       struct stat64 fname_stat;
--- a/localedata/Makefile
+++ b/localedata/Makefile
@@ -397,7 +397,7 @@ $(INSTALL-SUPPORTED-LOCALES): install-lo
 	@locale=`echo $@ | sed -e 's/^install-//'`; \
 	charset=`echo $$locale | sed -e 's,.*/,,'`; \
 	locale=`echo $$locale | sed -e 's,/[^/]*,,'`; \
-	flags="-c"; \
+	flags="-c --no-hard-links"; \
 	if [ "$$charset" = 'SHIFT_JIS' ] \
 	   || [ "$$charset" = 'SHIFT_JISX0213' ]; then \
 	   flags="$$flags --no-warnings=ascii"; \
